core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -516,6 +516,10 @@ object SparkSubmit extends CommandLineUtils {
     if (deployMode == CLIENT || isYarnCluster) {
       childMainClass = args.mainClass
       if (isUserJar(args.primaryResource)) {
+        val hadoopConf = new HadoopConfiguration()
+        args.primaryResource =
+          Option(args.primaryResource).map(
+            downloadFile(_, targetDir, args.sparkProperties, hadoopConf)).orNull
         childClasspath += args.primaryResource
       }
       if (args.jars != null) { childClasspath ++= args.jars.split(",") }
```

Comment:
For the feature completeness, we should guard out users who specify remote jars with `--jars` in cluster mode, your use case of HBase is just one special case, user could specify HBase dependencies with `--jars`, it is feasible. One the contrary, you can also specify a local `primaryResource` as a workaround.